patroni/log.py

```diff
@@ -4,63 +4,144 @@
 from copy import deepcopy
 from logging.handlers import RotatingFileHandler
 from patroni.utils import deep_compare
+from six.moves.queue import Queue
+from threading import Lock, Thread
 
+_LOGGER = logging.getLogger(__name__)
 
-class PatroniLogger(object):
+
+class QueueHandler(logging.Handler):
+
+    def __init__(self):
+        logging.Handler.__init__(self)
+        self.queue = Queue()
+        self._records_lost = 0
+
+    def _put_record(self, record):
+        self.format(record)
+        record.msg = record.message
+        record.args = None
+        record.exc_info = None
+        self.queue.put_nowait(record)
+
+    def _try_to_report_lost_records(self):
+        if self._records_lost:
+            try:
+                record = _LOGGER.makeRecord(_LOGGER.name, logging.WARNING, __file__, 0,
+                                            'QueueHandler has lost %s log records',
+                                            (self._records_lost,), None, 'emit')
+                self._put_record(record)
+                self._records_lost = 0
+            except Exception:
+                pass
+
+    def emit(self, record):
+        try:
+            self._put_record(record)
+            self._try_to_report_lost_records()
+        except Exception:
+            self._records_lost += 1
+
+    @property
+    def records_lost(self):
+        return self._records_lost
+
+
+class PatroniLogger(Thread):
 
     DEFAULT_LEVEL = 'INFO'
     DEFAULT_FORMAT = '%(asctime)s %(levelname)s: %(message)s'
 
     def __init__(self):
-        self.root_logger = logging.getLogger()
-        self.config = None
-        self.handler = None
+        super(PatroniLogger, self).__init__()
+        self.daemon = True
+        self._queue_handler = QueueHandler()
+        self._root_logger = logging.getLogger()
+        self._root_logger.addHandler(self._queue_handler)
+        self._config = None
+        self._log_handler = None
+        self._old_handlers = []
```

Comment:
are old log handlers closed separately to avoid extra locking in `reload_config`  or ?

Reply:
Exactly